web/source/kmwosk.ts

```diff
@@ -71,18 +90,37 @@ namespace com.keyman {
         var tId=((spec['text'] == '*Tab*' && spec.layer == 'shift') ? '*TabLeft*' : spec['text']);
 
         // Transforms our *___* special key codes into their corresponding PUA character codes for keyboard display.
-        t.innerHTML=this.renameSpecialKey(tId);
+        keyText=this.renameSpecialKey(tId);
       }
 
       //Override font spec if set for this key in the layout
       ts.fontSize=(<KeymanBase>window['keyman']).osk.fontSize;     //Build 344, KMEW-90
       if(typeof spec['font'] == 'string' && spec['font'] != '') {
         ts.fontFamily=spec['font'];
       }
+
       if(typeof spec['fontsize'] == 'string' && spec['fontsize'] != 0) {
         ts.fontSize=spec['fontsize'];
       }
 
+      let keyboardManager = (<KeymanBase>window['keyman']).keyboardManager;
+      var font = keyboardManager.getFont();
+
+      // Check the key's display width - does the key visualize well?
+      let width=OSKKey.getTextWidth(keyText, font);
+      if(width == 0 && keyText != '' && keyText != '\xa0') {
+        // Add the Unicode 'empty circle' as a base support for needy diacritics.
+        keyText = '\u25cc' + keyText;
```

Comment:
Yeah, the problem is even deeper -- sometimes it is font-specific as well ...  For example, try doing this with a Khmer vowel, first using a Windows font then with a KhmerOS font (from memory).

Reply:
Hmm, can you add a check for an existing \u25cc and \u200f, It looks like about 15 keyboards include 25cc, but I don't know if anyone else is using it the way I do. Otherwise, I guess I will need to change my KB when KM11 goes stable.

If Keyman starts to control the OSK font, and 25cc is rarely rendered in the text window (with the exception of Linguistics papers), there will be less issues, right? Maybe KM will one day need to include/download a fairly complete font, but I know that could almost double the Android download.

Reply:
The way this is written, an existing \u200f will prevent an extra one from being added, since \u200f will display with a width > 0.  (Unless neither display, in which case we have an entirely different issue.)  Note line 110 above - KeymanWeb will actively check whether or not your key cap text will display, only adding the extra character if and when it is needed.

The existing keyboards that you're referring to won't be affected.

Reply:
I see that now. Thx.